### PR TITLE
TMDM-9955 Give the possibility to launch the trigger only after transaction commit.

### PR DESCRIPTION
**Pull request by VictorLi-talendbj**
https://jira.talendforge.org/browse/TMDM-9955

**What is the current behavior?** (You should also link to an open issue here)

In Studio job, used tMDMConnection (not check 'Auto commit') and fill check box "Fire a Create/Update Event" on tMDMOutput, and use tMDMCommit or tMDMRollback to commit or rollback the update. but the triggers will be activated before the transaction commit.

**What is the new behavior?**

Update report event triggers will be executed after long transaction committed.

**Please check if the PR fulfills these requirements**

- [X] The commit message follows Talend standard
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [X] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.core/resources/META-INF/mdm-ehcache.xml [hunk 1]
```diff
@@ -18,4 +18,6 @@
          memoryStoreEvictionPolicy="LRU" />
     <cache name="basicAuthenticationCache" maxBytesLocalHeap="100M" eternal="false" timeToIdleSeconds="120" overflowToDisk="false" timeToLiveSeconds="120"
          memoryStoreEvictionPolicy="LRU" />
+    <cache name="updateReportEvents" maxBytesLocalHeap="100M" eternal="false" timeToIdleSeconds="300" overflowToDisk="false"
```
[the remark this answers is not in the document]
Comment (VictorLi-talendbj) [reply org.talend.mdm.core/resources/META-INF/mdm-ehcache.xml]: Done

File: org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java [hunk 2]
```diff
@@ -244,12 +267,71 @@ public void end(Committer committer) {
...
+    @Override
+    public void transactionCommitted(String longTransactionId) {
+        Object object = MDMEhCacheUtil.getCache(UPDATE_REPORT_EVENT_CACHE, longTransactionId);
+        if (object != null) {
+            List<String> stringObjects = (List<String>) object;
+            List<Document> updateReportList = new ArrayList<>(stringObjects.size());
+            for (String string : stringObjects) {
+                DOMDocument document = null;
+                try {
+                    StorageAdmin storageAdmin = ServerContext.INSTANCE.get().getStorageAdmin();
```
[the remark this answers is not in the document]
Comment (VictorLi-talendbj) [reply org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java]: Done

File: org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java [hunk 3]
```diff
@@ -244,12 +267,71 @@ public void end(Committer committer) {
...
+                    throw new RuntimeException("Failed to parse update report event.", e); //$NON-NLS-1$
                 }
+                updateReportList.add(document);
             }
+            routeItems(updateReportList);
         }
     }
 
+    @Override
+    public void transactionRollbacked(String longTransactionId) {
```
[the remark this answers is not in the document]
Comment (VictorLi-talendbj) [reply org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java]: Done

File: org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java [hunk 4]
```diff
@@ -53,6 +69,16 @@
 
     private boolean hasMetAutoIncrement = false;
 
+    private static TransactionService transactionService = new TransactionService();
+
+    public static final String UPDATE_REPORT_EVENT_CACHE = "updateReportEvents"; //$NON-NLS-1$
+
+    public static final String MDM_CACHE_MANAGER = "mdmCacheManager"; //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (VictorLi-talendbj) [reply org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java]: Done

File: org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java [hunk 5]
```diff
@@ -53,6 +69,16 @@
 
     private boolean hasMetAutoIncrement = false;
 
+    private static TransactionService transactionService = new TransactionService();
+
+    public static final String UPDATE_REPORT_EVENT_CACHE = "updateReportEvents"; //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (VictorLi-talendbj) [reply org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java]: Done

File: org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java [hunk 6]
```diff
@@ -244,12 +267,71 @@ public void end(Committer committer) {
...
+            }
+            saverSource.routeItem(dataCluster, typeName, itemsId);
+            iterator.remove();
+        }
+    }
+
+    private String getLongTransactionId() {
+        final TransactionManager transactionManager = ServerContext.INSTANCE.get().getTransactionManager();
+        Transaction currentTransaction = transactionManager.currentTransaction();
+        if (currentTransaction.getLifetime() == Lifetime.LONG) {
```
[the remark this answers is not in the document]
Comment (VictorLi-talendbj) [reply org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java]: Done

File: org.talend.mdm.core/src/com/amalto/core/storage/transaction/RouteItemListener.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                        throw new RuntimeException("Failed to parse update report event.", e); //$NON-NLS-1$
+                    }
+                    updateReportList.add(document);
+                }
+                SaverSession session = SaverSession.newSession();
+                session.routeItems(updateReportList);
+                MDMEhCacheUtil.removeCache(MDMEhCacheUtil.UPDATE_REPORT_EVENT_CACHE, longTransactionId);
+            }
+        } catch (Exception e) {
+            LOGGER.warn("Failed to route item as transaction committed.", e); //$NON-NLS-1$
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/transaction/RouteItemListener.java]: Should we use `LOGGER.error(xxx)`?
Should we call `MDMEhCacheUtil.removeCache` in `finally`?

File: org.talend.mdm.core/src/com/amalto/core/storage/transaction/RouteItemListener.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+                StorageAdmin storageAdmin = ServerContext.INSTANCE.get().getStorageAdmin();
+                Storage storage = storageAdmin.get(UpdateReportPOJO.DATA_CLUSTER, StorageType.MASTER);
+                MetadataRepository repository =  storage.getMetadataRepository();
+                ComplexTypeMetadata type = repository.getComplexType("Update"); //$NON-NLS-1$
+                for (String string : stringObjects) {
+                    DOMDocument document = null;
+                    try {
+                        document = new DOMDocument(Util.parse(string), type, UpdateReportPOJO.DATA_CLUSTER, UpdateReportPOJO.DATA_MODEL);
+                    } catch (Exception e) {
+                        throw new RuntimeException("Failed to parse update report event.", e); //$NON-NLS-1$
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/transaction/RouteItemListener.java]: `Failed to parse update report from cache.`

File: org.talend.mdm.core/src/com/amalto/core/storage/transaction/RouteItemListener.java (hunk 7 above)
Comment (VictorLi-talendbj) [reply org.talend.mdm.core/src/com/amalto/core/storage/transaction/RouteItemListener.java]: Done.

File: org.talend.mdm.core/src/com/amalto/core/storage/transaction/RouteItemListener.java (hunk 8 above)
Comment (VictorLi-talendbj) [reply org.talend.mdm.core/src/com/amalto/core/storage/transaction/RouteItemListener.java]: Done.

**Pull request merged**